src/ess/reflectometry/tools.py

```diff
@@ -108,3 +112,160 @@ def linlogspace(
+
+    All curves must be have the same unit for data and the Q-coordinate.
+
+    Parameters:
+        curves: the reflectivity curves that should be scaled together
+        return_scaling_factor:
+            If True the return value of the function
+            is a list of the scaling factors that should be applied.
+            If False (default) the function returns the scaled curves.
+
+    Returns:
+        A list of scaled reflectivity curves or a list of scaling factors.
+    '''
+    if len({c.data.unit for c in curves}) != 1:
+        raise ValueError('The reflectivity curves must have the same unit')
+    if len({c.coords['Q'].unit for c in curves}) != 1:
+        raise ValueError('The Q-coordinates must have the same unit for each curve')
+
+    qgrid = _create_qgrid_where_overlapping([c.coords['Q'] for c in curves])
+
+    r = _interpolate_on_qgrid(lambda c: c.data.values, curves, qgrid)
+    v = _interpolate_on_qgrid(lambda c: c.data.variances, curves, qgrid)
+
+    def cost(scaling_factors):
+        scaling_factors = np.concatenate([[1.0], scaling_factors])[:, None]
+        r_scaled = scaling_factors * r
+        v_scaled = scaling_factors**2 * v
+        v_scaled[v_scaled == 0] = np.nan
+        inv_v_scaled = 1 / v_scaled
+        r_avg = np.nansum(r_scaled * inv_v_scaled, axis=0) / np.nansum(
+            inv_v_scaled, axis=0
+        )
+        return np.nansum((r_scaled - r_avg) ** 2 * inv_v_scaled)
+
+    sol = opt.minimize(cost, [1.0] * (len(curves) - 1))
+    scaling_factors = (1.0, *sol.x)
+    if return_scaling_factors:
+        return [sc.scalar(x) for x in scaling_factors]
+    return [
+        scaling_factor * curve
+        for scaling_factor, curve in zip(scaling_factors, curves, strict=True)
+    ]
+
+
+def combine_curves(
+    curves: Sequence[sc.DataArray],
+    qgrid: sc.Variable | None = None,
+) -> sc.DataArray:
+    '''Combines the given curves by interpolating them
+    on a grid and merging them by the requested method.
+    The default method is a weighted mean where the weights
+    are proportional to the variances.
+
+    Unless the curves are already scaled correctly they might
+    need to be scaled using :func:`stitch_reflectivity_curves`.
+
+    All curves must be have the same unit for data and the Q-coordinate.
+
+    Parameters:
+        curves: the reflectivity curves that should be combined
+        qgrid: the Q-grid of the resulting combined reflectivity curve
+    Returns:
+        A data array representing the combined reflectivity curve
+    '''
+    if len({c.data.unit for c in curves}) != 1:
+        raise ValueError('The reflectivity curves must have the same unit')
+    if len({c.coords['Q'].unit for c in curves}) != 1:
+        raise ValueError('The Q-coordinates must have the same unit for each curve')
+
+    r = _interpolate_on_qgrid(lambda c: c.data.values, curves, qgrid)
+    v = _interpolate_on_qgrid(lambda c: c.data.variances, curves, qgrid)
```

Comment:
I don't know if interpolating the variances is an ok thing to do?
Whenever we interpolated anything, we typically just dropped the variances.
I can see that interpolating the variances could give you something reasonable, but maybe it should be made clear that from that point that the variances are no longer 'true' variances but something that was interpolated?

Above, when you compute the cost for the minimization, you do the same, but I think it's fine there because it's just to give us an idea of the contributions, internally.
Here, the variances will end up in the result that the user would pass on to analysis / publication, so I think we need to be careful?

Reply:
> I don't know if interpolating the variances is an ok thing to do?

I don't think it makes much difference, but I admit I haven't thought about it.

Here's one way to think about it: The interpolation is a linear combination of the two nearby points
```math
R_{interp} = a R_1 + (1-a) R_2
```
where $R_{interp}$ is the interpolated value, $a$ is the interpolation coefficient and $R_1$ respectively $R_2$ are the reflectivity values being interpolated.

The variance of the linear combination is
```math
Var[R_{interp}] = a^2 Var[R_1]  + 2 a (1-a) Cov[R_1, R_2]   +  (1-a)^2 Var[R_2] .
```
To compute the variance of the interpolated value we need some assumption about the covariance of the nearby $R$ values in the input data. 

What we currently do is to say that $Var[R_{interp}] = a Var[R_1]  +  (1-a)^2 Var[R_2] $.
That's the same as implicitly assuming
```math
 Cov[R_1, R_2]   = \frac{ Var[R_1] + Var[R_2] }{2}.
```
Don't know if that is the perfect assumption, but it's not a bad one. It's conservative because it assumes the neighboring values are positively correlated, and that the correlation is the same order of magnitude as the variances of the respective values.